src/sentry/metrics/minimetrics.py

```diff
@@ -148,61 +174,94 @@ def _hash(x: Any) -> int:
+            complexity_to_remove = 0
             buckets = self.buckets
+            force_flush = self._force_flush
+            flushed_buckets = set()
+            extracted_metrics = []
 
-            with self._lock:
-                for bucket_key, metric in buckets.items():
-                    ts, ty, name, unit, tags = bucket_key
-                    if ts > cutoff:
-                        continue
-
-                    m = {
-                        "timestamp": ts,
-                        "width": int(self.ROLLUP_IN_SECONDS),
-                        "name": name,
-                        "type": ty,
-                        "value": metric.serialize_value(),
-                    }
-                    if unit:
-                        m["unit"] = unit
-                    if tags:
-                        m["tags"] = dict(tags)
-
-                    extracted_metrics.append(m)
-                    cleanup.add(bucket_key)
-
-                for key in cleanup:
-                    buckets.pop(key)
-
-            if extracted_metrics:
-                self._emit(extracted_metrics)
-
-            time.sleep(2.0)
-
-    @staticmethod
-    def _emit(extracted_metrics: Any) -> Any:
-        # In order to avoid an infinite recursion for metrics, we want to use a thread local variable that will
-        # signal the downstream calls to only propagate the metric to the primary backend, otherwise if propagated to
-        # minimetrics, it will cause unbounded recursion.
-        thread_local.in_minimetrics = True
-
+            for bucket_key, metric in buckets.items():
+                ts, ty, name, unit, tags = bucket_key
+                if not force_flush and ts > cutoff:
+                    continue
+
+                extracted_metric = {
+                    "timestamp": ts,
+                    "width": int(self.ROLLUP_IN_SECONDS),
+                    "name": name,
+                    "type": ty,
+                    "value": metric.serialize_value(),
+                }
+                if unit:
+                    extracted_metric["unit"] = unit
+                if tags:
+                    # We need to be careful here, since we have a list of tuples where the first element of tuples
+                    # can be duplicated, thus converting to a dict will end up compressing and losing data.
+                    extracted_metric["tags"] = tags
+
+                extracted_metrics.append((extracted_metric, metric.current_complexity))
+                flushed_buckets.add(bucket_key)
+                complexity_to_remove += metric.current_complexity
+
+            # We remove all flushed buckets, in order to avoid memory leaks.
+            for bucket_key in flushed_buckets:
+                buckets.pop(bucket_key)
+
+            self._force_flush = False
+            self._bucket_complexity -= complexity_to_remove
+
+        if extracted_metrics:
+            self._emit(extracted_metrics, force_flush)
+
+    def kill(self):
```

Comment:
`kill` to me implies that it dies without flushing. This flushes once more, should be named `stop`.

Reply:
Perfect, thanks!